### PR TITLE
Adapt behavior to environment

### LINKED ISSUE #88: `activate()` should override `n` in `benchmark_run()`
In `benchmark_run(..., n = )`, `n` should default on a global R option that is changed on `activate()`. This simplifies local debugging.

```r
library(touchstone)
activate()
benchmark_run(myexpr = sys.sleep(4)) # should run once
````

So I propose: 
```r
benchmark_run <- function(expr_before_benchmark =
                            {},
                          ...,
                          branches = c(
                            branch_get_or_fail("GITHUB_BASE_REF"),
                            branch_get_or_fail("GITHUB_HEAD_REF")
                          ),
                          n = 100
                          path_pkg = ".") {
  # ...  override n with `'touchstone.n_runs'` if set.
}
```
`activate()` should set '`touchstone.n_runs'` to 1 (and deactivate should restore the previous value).

### LINKED ISSUE #89: Adapt behavior of `branch_get_or_fail()` to be environment depenent
``` r
setwd("~/git/touchstone/")
touchstone::branch_get_or_fail('GITHUB_BASE_REF')
#> ℹ touchstone not activated. Most likely, you want to run `touchstone::activate()` if you are working interactivley and the below error should go away.
#> Error: If you don't specify the argument `branch(s)`, you must set the
#> environment variable `GITHUB_BASE_REF` to tell touchstone which branches you
#> want to benchmark against each other.
#> ℹ See `?touchstone::run_script`.
```

<sup>Created on 2021-12-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

We can actually check if this is ran in interactive mode with `interactive()` before emitting this error. Because what we want to avoid is that people putting `activate()` in their `touchstone/script.R`. Also, we can check in `activate()` if it is ran in interactive mode and if not, emit at least a warning.

### LINKED ISSUE #94: Better GHA output
The gha log for touchstone benchmarks is not very informative due to repetition and length, I think there are ways to structure it better using output groups etc.. 

I'll looke into it.

### PR DESCRIPTION
**Pull request by assignUser**
Closes #88
Closes #89
Closes #94 

This will make debugging much quicker :D

### CODE REVIEW

**Comment by assignUser** [comment]
@lorenzwalthert I also updated the workflows in this repo to use the new actions

**Comment by lorenzwalthert** [comment]
Thanks. Will habe a look later. If this package passes tue linking check, I suggest you activate the corresponding pre-commit hook too in `.pre-commit-config.yaml`.

File: R/utils.R [hunk 1]
```diff
@@ -371,3 +372,9 @@ get_git_root <- function() {
 
   git_root
 }
+
+gh_cat <- function(string) {
+  if (Sys.getenv("GITHUB_ACTIONS") == TRUE) {
```
Comment (lorenzwalthert) [inline R/utils.R:377]: You don't need the equality comparison if the condition is `TRUE`. Also, I think these are always strings 

```r
Sys.setenv("GITHUB_ACTIONS" = TRUE)
Sys.getenv("GITHUB_ACTIONS")
#> [1] "TRUE"
```
So I suggest you do `as.logical(upper(Sys.getenv("GITHUB_ACTIONS")))`.

File: .lintr [hunk 2]
```diff
@@ -0,0 +1,2 @@
+linters: with_defaults(line_length_linter(120), open_curly_linter = NULL)
+exclusions: list("inst/script.R","R/zzz.R"=1)
```
Comment (lorenzwalthert) [inline .lintr:2]: What does `=1` mean? Also, should we exclude whole files or use the ` "# nolint"` or ` "# nolint start"` markers?

Comment (assignUser) [reply .lintr:2]: `file.R=<linenumber>` so just exclude line 1, in this case due to on.load breaking snake_caee convention

File: R/utils.R (hunk 1 above)
Comment (assignUser) [reply R/utils.R:377]: Yeah being explicit might be better than relying on implicit casting...

File: R/source.R [hunk 3]
```diff
@@ -63,12 +65,14 @@ run_script <- function(path = "touchstone/script.R",
...
-#'   to merge your changes into) when running benchmarks.
-#'   Defaults to 'main' if the option `touchstone.default_base_branch`is not set.
+#' @param head_branch Git branch to be used as the `GITHUB_HEAD_REF` branch
+#' (i.e. the branch with new changes) when running benchmarks. Defaults to the
+#' current branch.
+#' @param base_branch Git branch for the `GITHUB_BASE_REF` (i.e. the branch you
+#' want to merge your changes into) when running benchmarks. Defaults to 'main'
+#' if the option `touchstone.default_base_branch`is not set.
+#' @param n Number of times benchmarks should be run for each `branch`. Will
+#' override `n` argument in all interactive calls to [benchmark_run]
```
Comment (lorenzwalthert) [inline R/source.R:75]: I think tidyverse style compliant {roxygen2} is indented on line break, i.e. 
```r
#' @param x this is param x and the
#'   description does not fit on one line.
NULL
```
Can you adapt it?

File: R/source.R [hunk 4]
```diff
@@ -83,11 +87,28 @@ activate <- function(head_branch = gert::git_branch(),
...
+      ))
+    }
+  }
+
+
   suppressMessages({
     withr::local_envvar(
       GITHUB_BASE_REF = base_branch,
       GITHUB_HEAD_REF = head_branch,
+      touchstone.n_runs = n,
```
Comment (lorenzwalthert) [inline R/source.R:111]: Just to remain consistent here, I think this should be an R option, not an env variable.

File: R/source.R [hunk 5]
```diff
@@ -83,11 +87,28 @@ activate <- function(head_branch = gert::git_branch(),
                        "touchstone.default_base_branch",
                        "main"
                      ),
+                     n = 1,
                      env = parent.frame()) {
+  if (!rlang::is_interactive()) {
+    if (Sys.getenv("GITHUB_ACTIONS") == TRUE) {
```
Comment (lorenzwalthert) [inline R/source.R:93]: Same as for the other condition where you have `== TRUE`.

File: R/source.R (hunk 3 above)
Comment (lorenzwalthert) [reply R/source.R:75]: Also, must end with full stop, and to get code font formatting, we need `[benchmark_run()]`, not `[benchmark_run]` I think.

File: R/core.R [hunk 6]
```diff
@@ -95,9 +96,15 @@ benchmark_run <- function(expr_before_benchmark =
     abort_string()
   }
 
+  if (rlang::is_interactive()) {
+    n <- Sys.getenv("touchstone.n_runs", unset = NA) %>% as.numeric() %|% n
```
Comment (lorenzwalthert) [inline R/core.R:100]: I think to follow the *repeat yourself* paradigm in documentation, we should emit a message here that n is overridden and also in the documentation above for `n`, we could make this explicit,

**Comment by assignUser** [comment]
Agree on the changes, will implement them.

File: R/source.R (hunk 3 above)
Comment (assignUser) [reply R/source.R:75]: mhm roxygen formatting is always a pain... I wish there was a package that could *style* those comments. :wink:

**Comment by assignUser** [comment]
@lorenzwalthert Done :) 
I disabled the `object_usage_linter` as I only ever get false positves (e.g. when vars are only used in `glue`) and don't think `# nolint`ing every false positv is worth it, what do you think?

File: R/source.R [hunk 7]
```diff
@@ -30,9 +32,10 @@
 #' contains the installed benchmarked package for set-up tasks, and temporarily
 #' remove it during benchmarking with [benchmark_run()] so only one
 #' touchstone library is on the library path at any time.
+#'
 #' @return
-#' The same as [base::source()], which inherits from [base::withVisible()], i.e.
-#' a list with `value` and `visible` (invisibly).
+#'   The same as [base::source()], which inherits from [base::withVisible()], i.e.
```
Comment (lorenzwalthert) [inline R/source.R:37]:
```suggestion
#' The same as [base::source()], which inherits from [base::withVisible()], i.e.
```

File: R/source.R [hunk 8]
```diff
@@ -30,9 +32,10 @@
...
+#'   a list with `value` and `visible` (invisibly).
```
Comment (lorenzwalthert) [inline R/source.R:38]:
```suggestion
#' a list with `value` and `visible` (invisibly).
```

File: R/source.R (hunk 7 above)
Comment (lorenzwalthert) [reply R/source.R]: So if you use 
```r
#' @param x this is 
#'   multi line
```

and 
```r
#' @section x
#' Here no indention.
```

As specified [here](https://style.tidyverse.org/documentation.html#indents-and-line-breaks).

**Review by lorenzwalthert — APPROVED** [review]

**Comment by lorenzwalthert** [comment]
Thanks @assignUser 👍

**Pull request merged**